### ISSUE TITLE
Unable to get Node Express & React app working with csrf-csrf

### ISSUE THREAD

**Issue by kashaf-s** [Psifi-Solutions/csrf-csrf#54]
Hey, I fairly new to the CSRF concept but security team has asked us to do an anti-CSRF implementation for our app and since the `csurf` package was deprecated, I turned to this package. But I'm having a very hard time implementing this.All the methods I have tried return a 403 from the frontend. I am using node express as a backend and React as frontend - using axios to make the api calls.

Here's my latest implementation - 
Server.js -
```
const bodyParser = require('body-parser');

const { doubleCsrf } = require("csrf-csrf");
const cookieParser = require("cookie-parser");
const router = require("./app/router");
const app = express();
const corsOptions = {
  credentials: true,
  origin: allowOrigin,
  optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
};

app.use(cors(corsOptions));
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: ["'self'"],
        scriptSrc: ["'self'", "'unsafe-inline'", "'unsafe-eval'"],
        styleSrc: ["'self'", "https:", "'unsafe-inline'"],
        baseUri: ["'self'"],
        fontSrc: ["'self'", "https:", "data:"],
      },
    },
  })
);
app.use(cookieParser());

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

const {
  generateToken, // Use this in your routes to provide a CSRF hash + token cookie and token.
  doubleCsrfProtection, // This is the default CSRF protection middleware.
} = doubleCsrf({
  cookieName: "_csrf",
  cookieOptions: {
    secure: true,
  },
  getSecret: () => "supersecret",
  getTokenFromRequest: (req) => req.headers["_csrf"],
});
app.use(doubleCsrfProtection);
app.use((req, res, next) => {
  res.locals.csrfToken = generateToken(req, res);
  next();
});
// register routes
app.use("/api", router);
....
```

Before trying this, I also tried creating my own middleware and an api end point that would generate the token and the frontend would send it back on the actual put request - 
In my route middleware - 
```
class RouteMiddleware {
  constructor() {
    const CSRF_COOKIE_NAME = 'csrf-token'; // or 'Csrf-Token'
    const { doubleCsrfProtection, generateToken } = doubleCsrf({
      cookieName: CSRF_COOKIE_NAME,
      cookieOptions: { sameSite: false, secure: false, signed: true },
    });

    this.csrfProtection = doubleCsrfProtection;
    this.generateToken = generateToken;
  }

  generateCsrfToken(req, res) {
    const csrfToken = this.generateToken(req, res);
    return csrfToken;
  }
}
```

In my router.js - 
```
const express = require('express');
// import helpers
const { routeMiddleware } = require('./lib');
const router = express.Router();
...
router.get('/csrf-token', (req, res) => {
  const csrfToken = routeMiddleware.generateCsrfToken(req, res);
  res.json({ token: csrfToken }); 
});
...
router.put('/user/profile', routeMiddleware.csrfProtection, userController.update);
...
```

And in the frontend code: 
```
generateCsrfToken().then((csrf) => {
      const formSubmitData = {
        userId: user._id,
        email: user.email,
        csrfToken: 'csrf',
      }
      // this function makes a call to the protected route - /user/profile
      userUpdateFn.mutate(formSubmitData, { onSuccess })
    }
```

Both these implementations return 403 forbidden errors from the frontend, even when I can see the token being passed correctly -
Response from /csrf-token -
```
{
    "token": "13ec99ac68ff980503e66ffa47243b6e8889b4ddd3ce7437aaa229ad65afdc312b45df9bf0b1ad1e20b4d55a2660bcec5df17a8844c3fc5cf876acf84c03d360"
}
```

headers in the protected route (/user/profile) - (i tried changing the service to read capital-case too, but it didn't work )
![Screenshot 2024-02-13 at 12 03 45 pm](https://github.com/Psifi-Solutions/csrf-csrf/assets/93901860/629e08e3-4af1-435b-b310-4ef9b3526e90)

**Comment by psibean** [comment]
There are some "Change for development" sections in the readme. I really should just have a "Setting up for local development" section, as the default configuration is intended for production. But there are also some examples in the repository itself, albeit there could be better ones.

> secure: true

Unless you're specifically running your app with SSL on your local machine, this won't work. You'll typically want to use some kind of environment variable to determine this value, e.g.:

```
const IS_PRODUCTION = process.env.APP_ENVIRONMENT === 'production'; // where you set APP_ENVIRONMENT as an environment variable in your live environment
```

and then 

```js
secure: IS_PRODUCTION
```

Means this will be set to false when on your local machine.

Additionally, depending on your app, you may need to change the cookie name and remove the secure `__Host-` prefix, this depends on whether you're using SSR and / or whether your frontend is served on the same domain as the backend service. In local development, this likely isn't the case, and it may vary depending on your deployment setup.

**Comment by chr15m** [comment]
@psibean I don't think that's correct. Secure cookies are sent to localhost. This is a quote from MDN:

> A cookie with the Secure attribute is only sent to the server with an encrypted request over the HTTPS protocol. It's never sent with unsecured HTTP (**except on localhost**)

**Comment by chr15m** [comment]
@kashaf-s I believe the issue is as follows (@psibean can correct me if this is wrong):

You have the setting `cookieName: "_csrf"` and so csrf-csrf is setting the cookie.

However your `getTokenFromRequest` is reading the value from `headers` not from cookies:

```
  getTokenFromRequest: (req) => req.headers["_csrf"],
```

@psibean is it save to read from (secure, httpOnly) cookies as follows, assuming you have cookies middleware set up?

```
  getTokenFromRequest: (req) => req.cookies["_csrf"].split("|")[0],
```

**Comment by chr15m** [comment]
I also found integration of this module confusing. It would be great if there was a "replacing default csurf usage" example for people coming from that old module.

**Comment by chr15m** [comment]
@psibean ok understood. Thanks for schooling me on this and helping me avoid disaster! 🙏

**Comment by psibean** [comment]
Hey @kashaf-s I can tell from your original post that my last reply (specifically the edit part) should resolve the problem you're facing here. I'll repeat it here just in case:

>  In your screenshot it looks like your header name is Csrf-Token, is that correct? If so, then you should set this accordingly:

```
getTokenFromRequest: req => req.headers['Csrf-Token']
```

> or change your frontend header name to _csrf

I'll close this for now - feel free to re-open if this doesn't solve the problem!

**Comment by psibean** [comment]
Hey @kashaf-s I can tell from your original post that my last reply (specifically the edit part) should resolve the problem you're facing here. I'll repeat it here just in case:

>  In your screenshot it looks like your header name is Csrf-Token, is that correct? If so, then you should set this accordingly:

```
getTokenFromRequest: req => req.headers['Csrf-Token']
```

> or change your frontend header name to _csrf

I'll close this for now - feel free to re-open if this doesn't solve the problem!

**Comment by LouisVA** [comment]
As a side note, I see @kashaf-s doing

app.use(doubleCsrfProtection);
app.use((req, res, next) => {
  res.locals.csrfToken = generateToken(req, res);
  next();
});

while the documentation says: 
"Just make sure you register this route before registering the middleware so you don't block yourself from getting a token."
express.use(session);
express.get("/csrf-token", myRoute);
express.use(doubleCsrfProtection);

with the myRoute containing the generateToken method.

Is OP blocking himself from getting a token by calling doubleCsrfProtection before the route that generates it?

**Comment by psibean** [comment]
> while the documentation says: "Just make sure you register this route before registering the middleware so you don't block yourself from getting a token." express.use(session); express.get("/csrf-token", myRoute); 

Oop, that documentation is wrong. 🤦🏻 By default, GET requests are not protected, because they don't usually need to be. GET requests should not be making any kind of server state mutation, they should not be making any writes - GET requests should only ever be used for read based actions. And since CSRF attacks don't have access to response content, targeting GET requests achieves nothing.

**Comment by psibean** [comment]
> @psibean Good point, I was aware it was not blocking GET but that didn't pop-up when reading this part of the documentation.

Yep, the docs are wrong. Originally get requests weren't ignored, when they should have been. Looks like I missed updating the doc at the same time. Ideally, the route in that case should come after the middleware, because then that gives you access to `req.csrfToken` alias.

> I am myself struggling with the implementation. I already debugged a bit and saw that my backend is receiving the correct csrf token in the header but can't find the token/hash in the cookies!

The cookie should be sent automatically by the browser with any request sent to the domain the cookie was sent on. Depending on your [same-site](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value) setting, this may be restricted. If you need the cookie to be sent from domains different to the one that set it, you'll need to set the same-site to `Lax` (in the case of re-direct based requests), or `none`.

If you can't have same-site set to `strict` **it is highly recommended** that you have your CORS configuration set with an explicit list of domains (allowed origins) which you're permitting cross-site requests to come from. This can typically be configured on the application level via the cors middleware, or on your web server level (e.g. nginx or apache config).

> Basically users get forwarded to Keycloak in the frontend to authenticate, then the backend finishes the oauth authentication and saves the keycloak access tokens for the user.

For the OAuth2 workflow, you'll typically want to use the authorization code grant flow and supply the `state` query parameter. With OAuth2, the `state` parameter is used to avoid CSRF attacks. In the [latest RFC draft by the  Internet Engineering Task Force (IETF) for OAuth 2.0 Security Best Practices](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics) - use of the state parameter for this purpose is recommended.

The flow should typically look like this:

1. Your user gets re-directed to the OAuth2 login. The URL will have the code type, the state parameter, a redirect uri and typically a defined scope and clientid.
2. The user logs in via the OAuth2 provider, this provider then makes a redirect to the provided callback URL, this URL contains the granted code and the state parameter. You can at this point re-direct to a page on your main domains frontend, that frontend page will read the code grant and state parameters and then post them to your backend. Otherwise you can also re-direct directly to the backend, which will then still receive the query parameters.
3. Your backend receives the code grant and state parameter, it verifies the state parameter and then uses the code grant to request the access token.
4. You use the access token to request the identity, or in the case of OICD, the token should have the necessary claims.

There shouldn't be any part of this process where the OAuth2 provider needs to send a CSRF token back to the first-party domain.

> I was not using cookieparser before, so just added it for the csrf-csrf. I understand that maybe the sync version of the protocol might fit better. But I am not really keeping state, the backend is REST

You are keeping state. You're using `express-session` which by default keeps state, that's the entire point of these kinds of sessions. You've even configured it with a store for storing state. `csrf-sync` works quite well with a properly configured session like this. 

> // some auth routes then that include a POST, so I don't want these to be blocked by CSRF

Concerned with this. CSRF protection doesn't block anything, it protects it. If it's blocking these requests, then you've done something wrong. Auth routes should absolutely be protected by CSRF. If these routes are intended to received request from different domain/origin, then as mentioned above, you may need to adjust the cookie configuration so those requests include the cookie.

**If you need to reply to this post with additional questions, please do so in a new issue.**

**Comment by kashaf-s** [comment]
Hey guys, really appreciate the help. I will be trying out the things mentioned here in the next week, and will let you know what's working. We have an okta integration so assuming it handles csrf, but we were told by our Penetration testing teams that it would be nice to have an added layer of security by having our own anti-csrf implementation on top of it.
